### ISSUE TITLE
How to Configure the Faban threadstart Option

### ISSUE THREAD

**Issue by VincenzoFerme** [benchflow/benchflow#147]
_From @VincenzoFerme on May 16, 2016 8:18_

Faban enables the following configuration:

``` xml
<threadstart>
<!-- The avg time between the start of each thread in
an agent -->
<delay>1000</delay>
<!-- Whether the agents will wait until all threads to
be up and then start the load all at once. If
set to false, each thread will start executing the
operations as soon as it is started. This will
gradually load up the server. -->
<simultaneous>false</simultaneous>
<!-- Whether the agents will start up threads in parallel
or not. If set to false, an agent will start all it's
threads before the next agent gets to start the
threads. The socket connection queue will be less
loaded when it is set to false. -->
<parallel>false</parallel>
</threadstart>
```

`delay` impacts on the velocity by which the drivers get started, especially if `parallel` is set to `false`. This happens because with sequential start, a new thread starts only after `delay` milliseconds. This behaviour has an impact on the actual `rampUp` time. In the following we report how to determine the value for the `delay` configuration:

``` java
//the default (and minimum) value should be 200 in case the rampUp is set to 0, as [defined by Faban](https://github.com/akara/faban/blob/d6832a3833dd0ce0d713d7fa7f178ea24e77d605/driver/src/com/sun/faban/driver/engine/RunInfo.java#L99). We currently set 0 and test if it works correctly.
if rampUp == 0
   delay = 0
else if parallel == false
   delay = (rampUp/scale) * 1000 //we multiply * 1000 because the rampUp time is in seconds, and the delay time is in milliseconds
else if parallel == true
   delay = (rampUp/scale) * #(agents+master utilised) * 1000 //we multiply * 1000 
```

The values should be round to the `half to even` because we get a thread starting time close to the `rampUp` and the Faban anyway waits for the `rampUp` time to complete if we get a time less that `rampUp`. The alternative would be to round to the `floor`, but in that case we can get all the thread starting issuing the load during the `rampUp` for a while.

`simultaneous` is set to `false` by default so that we don't overload the system under test. There will be certain type of test, where we need to change this value to `true`, so make it configurable according to the type of test (a concept that should be in the system, and it's default is `custom`)

`parallel` should be set to `true` by default and configurable. Currently we don't see the necessity to reduce the socket connection queue load. Setting it to true make the agents starting process faster.

_Copied from original issue: benchflow/drivers-maker#36_

**Comment by VincenzoFerme** [comment]
_From @simonedavico on June 9, 2016 10:3_

@VincenzoFerme question about the case in which `parallel == true`, in particular about the meaning of `#(agents + master utilised)`: what does it mean? Number of agents utilised in total? Or Number of hosts + master in total? Or something else entirely?

**Comment by VincenzoFerme** [comment]
They are the number of agents you instantiate, according to https://github.com/benchflow/drivers-maker/issues/32, plus one for the master, if you end up using the master host to allocate agents to.

**Comment by VincenzoFerme** [comment]
_From @simonedavico on August 19, 2016 15:53_

Example, for the sake of comprehension of the algorithm:

Assume we have 2 agents processes and 5 threads for each.

If we set `parallel` to `false`, the threads will be started sequentially: 1,2,3...10.
Thus, the delay between each has to be `rampUp/#threads`. Since the number of threads is the number of simulated users configured for the experiment, we can compute `rampUp/users`.
In this case it would be `1 (* 1000)`.

On the other hand, if we set `parallel` to `true`, both agent processes will start their threads together, thus we will have: `(1,2,3,4,5) // (1,2,3,4,5)`. In this case, the delay has to be the `rampUp` divided by the number of threads started for each agent, i.e., `rampUp/users * #agentProcesses`. In this case it would be `2 (* 1000)`.

Note that in the [original explanation](https://github.com/benchflow/drivers-maker/issues/36#issue-154982142) `scale` actually refers to the number of simulated users.
